### PR TITLE
is-shallow-equal: Convert to ESM

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Description
This PR converts `is-shallow-equal` to ES modules. I'm proposing this in order to allow it to be consumed in an ESM context. ES modules have matured in node to the point where this should be viable.

## How has this been tested?
Original tests continue to pass. Build passes. Benchmark runs.

## Types of changes
Breaking change. See https://github.com/webpack/webpack/issues/4742 for details on how importing this module in CJS context needs to change. `benchmark/index.js` in this changeset is a good example of how this will change.

I actually noticed that the benchmark seems to be slower than the README. This is true on `master` and this branch. Should I also update the benchmarks? Specifically these can be slower (on `master`):

```
@wordpress/is-shallow-equal (object, equal) x 4,933,251 ops/sec ±0.76% (88 runs sampled)
@wordpress/is-shallow-equal (object, same) x 196,122,526 ops/sec ±1.04% (88 runs sampled)
@wordpress/is-shallow-equal (object, unequal) x 5,450,480 ops/sec ±1.03% (85 runs sampled)
@wordpress/is-shallow-equal (array, equal) x 26,996,443 ops/sec ±1.01% (88 runs sampled)
@wordpress/is-shallow-equal (array, same) x 61,463,049 ops/sec ±0.96% (89 runs sampled)
@wordpress/is-shallow-equal (array, unequal) x 28,314,907 ops/sec ±0.79% (91 runs sampled)
```

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by gziolo** [comment]
At some point, we converted this package to ESM but we reverted it back to ES5 because of some breaking changes introduced. See: https://github.com/WordPress/gutenberg/pull/8132.

It shouldn't be an issue anymore, now that https://github.com/WordPress/gutenberg/pull/18942 landed with a new major version that removed this limitation.

Some tools might still assume that `is-shallow-equal` is ES5 based, so we will have to update that as well. Although, I couldn't find any place that would be affected. It looks like we made configuration for tools more flexible :)

**Comment by gziolo** [comment]
I checked 2 failing E2E jobs for mobile apps and it looks like those are valid issues to resolve.

**Review by hc0503 — APPROVED** [review]

**Comment by saramarcondes** [comment]
@gziolo Any ideas how to address this? At first I thought it was merely a resolution issue as the original error stated something along the lines of:

```
Unable to resolve ".../index.js(index.js|.js|...)"
```

I assumed the issue here was a double `.js`.

I added a `react-native` entry to the package.json to match other packages I observed doing the same thing, but that did not resolve the issue 🤔 Any help would be appreciated!

File: packages/is-shallow-equal/package.json [hunk 1]
```diff
@@ -25,7 +25,9 @@
 		"build-types",
 		"*.md"
 	],
-	"main": "lib/index.js",
+	"main": "build/index.js",
+	"module": "build-module/index.js",
+	"react-native": "build/index",
```
Comment (sirreal) [inline packages/is-shallow-equal/package.json:30]: Should this refer to `src/index` like e.g. a11y does?

https://github.com/WordPress/gutenberg/blob/2296afb65651f67d132e3c770c9372653e1575dc/packages/a11y/package.json#L25

```suggestion
	"react-native": "src/index",
```

File: packages/is-shallow-equal/src/objects.js [hunk 2]
```diff
@@ -1,6 +1,4 @@
-'use strict';
-
-var keys = Object.keys;
+const keys = Object.keys;
```
Comment (sirreal) [inline packages/is-shallow-equal/src/objects.js:1]: `keys` appears to be used twice, stylistic but I'd prefer to just write `Object.keys` twice 🙂

**Review by sirreal — CHANGES_REQUESTED** [review]
`files` in package.json needs to be updated:

https://github.com/WordPress/gutenberg/blob/bb3f0b4ee0c4673a940ab089ee886770c88b99ce/packages/is-shallow-equal/package.json#L23-L27

These are pesky because you don't catch them until you've published a broken package. I know from experience 😅 

Maybe remove it entirely?

**Comment by saramarcondes** [comment]
@sirreal Good catch! I've updated it. There's no reason to distribute the benchmark and test files I don't believe.

**Comment by sirreal** [comment]
Package lock problems now, maybe a rebase 😕

**Comment by gziolo** [comment]
I tried to run benchmarks as noted in the README file
 (https://github.com/WordPress/gutenberg/blob/master/packages/is-shallow-equal/README.md#benchmarks)

It needs to be updated to include the step that builds packages:

```bash
cd packages/packages/is-shallow-equal
npm install
npm run build:packages
node benchmark
```

You might see the following issue when you run the benchmark for the first time:

```bash
(node:65044) UnhandledPromiseRejectionWarning: Error: Cannot find module '@wordpress/lazy-import.5ab856344c555e50779f56b4cd1f2e02'
```

It's a known issue. All subsequent runs should work as expected.

There is an issue with importing code using CJS:

https://github.com/WordPress/gutenberg/blob/master/packages/is-shallow-equal/benchmark/index.js#L43-L44

This code probably errors as I no longer see benchmark results:
```bash
@wordpress/is-shallow-equal (object, equal): 
@wordpress/is-shallow-equal (object, same): 
@wordpress/is-shallow-equal (object, unequal): 
@wordpress/is-shallow-equal (array, equal): 
@wordpress/is-shallow-equal (array, same): 
@wordpress/is-shallow-equal (array, unequal):
```

**Review by gziolo — CHANGES_REQUESTED** [review]
See my previous comment, otherwise, it's good to go.

**Comment by saramarcondes** [comment]
@gziolo Thanks for catching that. I've fixed the issue and noted that there is now a breaking change (really glad you tried out the benchmark).

I also updated the description to note that the benchmarks are different from the README's. This is true of `master` as of this branch (though there is no significant difference between this branch and `master`). Shall I also update the README with new benchmarks?

File: packages/is-shallow-equal/CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,12 @@
 
 ## Unreleased
 
+## 3.0.0 (2020-11-12)
```
Comment (gziolo) [inline packages/is-shallow-equal/CHANGELOG.md:5]: Release tool will add a version number and date. It's all automated based on type of sections listed.

File: packages/is-shallow-equal/package.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
 	"name": "@wordpress/is-shallow-equal",
-	"version": "2.3.0",
+	"version": "3.0.0",
```
Comment (gziolo) [inline packages/is-shallow-equal/package.json:3]: Here as well, it should remain unchanged. Lerna will bump version at the time of npm publishing. 

How does it work in Calypso, I'm curious 😃

**Review by gziolo — COMMENTED** [review]
It would be nice to update benchmark's results. It might be a follow up PR if other libraries listed are outdated.

File: packages/is-shallow-equal/package.json (hunk 4 above)
Comment (saramarcondes) [reply packages/is-shallow-equal/package.json]: Ah I should have read the link at the top of the file 😉 In Calypso these are maintained manually as far as I know!

**Review by gziolo — APPROVED** [review]
It is good to go, thank you 😃

**Comment by saramarcondes** [comment]
@sirreal Can you re-review this PR? It's blocked on your change request (which I addressed 🙂)

**Review by sirreal — APPROVED** [review]
Looks good, thanks.

**Pull request merged**